Add unique tag per Summarizer run to Summary logs

markfields opened the pull request:
Fixes https://github.com/microsoft/FluidFramework/issues/4018

This adds a `summarizerRunTag` property to many summarizer-related logs which is a counter that increments whenever a new summary is kicked off by `RunningSummarizer`. It's unique within a `Summarizer` instance which is 1-to-1 with the container so in the logs look for `containerId`.

I also did some refactoring around the seam between container runtime and summarizer, which I think could use more work in the future to achieve better separation of concerns.

markfields committed:
Add SummaryId property to Summarizer logs

Co-authored-by: Kabir Brar <kabrar@microsoft.com>

markfields committed:
lock

markfields commented:
Kabir and I were seeing some ID traces start with the SummarizerStop event. So this approach may be too naive, need help to understand logical and code-wise lifecycle of Summarizers.  cc @arinwt

ksbrar commented on packages/runtime/container-runtime/src/summarizer.ts:
```diff
@@ -598,7 +599,7 @@ export class Summarizer extends EventEmitter implements ISummarizer {
     ) {
         super();
-        this.logger = ChildLogger.create(this.runtime.logger, "Summarizer");
+        this.logger = ChildLogger.create(this.runtime.logger, "Summarizer", { SummaryId: uuid() });
```
SummarizerId vs. SummaryId? (Not sure myself).

vladsud replied:
SummarizerInstanceId ?
Though looking at container.ts, we use 
                containerId: uuid(),
So I guess it's SummarizerId.

anthony-murphy replied:
i dislike containerId, and in fact bohemia overwrites it with their own value is some cases. I'd rather not use id, as i think it is misleading, maybe use the term tag instead, so SummarizerTag

curtisman replied:
If we are switching to `Tag`, we should be consistent in other places though.

markfields replied:
Turns out this guid is 1-1 with the container's logging ID. Not very useful. Working on the counter of summaries in a container like Tony suggested - it already exists, just isn't stamped on all the events you'd want it to be.

anthony-murphy commented:
I like the Summarizer level correlation. Do we also want a per summary correlation?

vladsud commented:
it does not hurt to add it if it's simple. The only downside is telemetry cost - we actually generate a lot of data, and once we go to GA, it will be pricey. Once all IDs settle, we should look into Area reports - they have pretty good tables describing storage cost (accounting for compression)

anthony-murphy commented:
good point. rather than a new correlation maybe just the seq? that would server dual function then

markfields commented:
Thanks for the input everyone. @arinwt and I sat together for 2 hours this afternoon looking at how to properly stamp a simple summary counter onto all the relevant log lines.  We're looking at [this trace](https://kusto.aria.microsoft.com/6a8929bcfc6d44e9b13fee392ada9cf0?query=H4sIAAAAAAAEAI1QTU8CMRC9k%2fAfJnveTUBklQMHP9CYGCBC4nHTbadQQ9t12mog%2fngLuxAl%2bHGZzLx5fW9eJ1IqjsXdKihR16dgvNJY3KNBUrzd%2boBglDVga6rcUXeVGurimDr5UXWKJC1pZjj%2bhz4isrQlvi%2bREG6ZZwW3xjMVHR8EDIeQiG63lNjrZINOPsjOe4N%2bVgqeZ5fI8p44u%2bjmfUyONPANjR8zjbBkDpJZ0JqR2iAlYOkXyvqPvdoos3hWfnnNHE5XwU0qt%2fOuyL4g9zDavirmMVt6JJMCtzE1DGvc7U%2b62aIpOCbxsIt9Gm2NWB2wekqBUO6h2M7wNWD863HQJVJj6U6BtnIzFcFH5nyT9cTmynvUld8LhVKriIhm1ugcW2D6NcH6%2bwntVrv1CYeFUq90AgAA&web=0

Office_Fluid_FluidRuntime_Generic
| union office_fluid_fluidruntime_generic
| union Office_Fluid_FluidRuntime_Performance
| union Office_Fluid_FluidRuntime_Error
| where Data_containerId == "d11bfe30-9069-4395-bdc6-8ea63d27165e"
| where Data_eventName has "Summarizer" or Data_eventName has "Summary" or Data_eventName has "SummarizingWithBasePlusOps"
| project Event_Time, Data_eventName, count = Data_summarizeCount, safe = Data_safe, handle = Data_handle, ref = Data_refSequenceNumber, Data_sequenceNumber, Data_opsSinceLastSummary, Data_opsSinceLastAttempt, Data_submitted, Data_message, Data_summarySequenceNumber

) as an example, which has logs from the RunningSummarizer (fluid:telemetry:Summarizer:Summarizing_start), SummarizerNodes (fluid:telemetry:SummarizingWithBasePlusOps), and the Summarizer class that listens to acks (fluid:telemetry:Summarizer:SummaryAck).

markfields committed:
Pass around a summaryLogger

This is a ChildLogger with namespace "Summary"
and property "summaryCount" which is
the sequence number of this summary
since the summarizer instantiated.

markfields committed:
Revert guid on Summarizer's logger

And a small rename

markfields commented:
Converted to a draft because I haven't done any testing (or written any tests... if that's doable/worthwhile for logging), and it's a bit gnarly all the plumbing in here. Might consider if there's some mild refactoring that clears up what's going on here.

Also, there is storage code and I'm sure other code that's dispatched in the midst of a summary, and for now I'm not plumbing the summaryLogger there. ContainerId and the event time sequence is enough to paint the picture when looking in detail. If we find ourselves wanting to aggregate all those logs by summary counter... we can reconsider.

vladsud commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -877,8 +877,10 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
-            async (full: boolean, safe: boolean) => this.generateSummary(full, safe),
-            async (propHandle, ackHandle, refSeq) => this.refreshLatestSummaryAck(propHandle, ackHandle, refSeq),
+            // eslint-disable-next-line max-len
+            async (full: boolean, safe: boolean, summaryLogger: ITelemetryLogger) => this.generateSummary(full, safe, summaryLogger),
```
this.generateSummary.bind(this) is better IMHO, same for another one

arinwt replied:
Why do you prefer `bind` over arrow function?

Not sure if this has much weight, but I found [this style guide.](https://google.github.io/styleguide/jsguide.html#features-functions-arrow-functions)

If it's just too lengthy and redundant to type out all the individual args on both sides, could we just do something like?
```
async (...args) => this.refreshLatestSummaryAck(...args)
```

vladsud replied:
Too verbose and need to change every time args are changed. ...args would work too

markfields replied:
I like ...args.  I'll clean these up on Monday.

vladsud commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -2155,6 +2158,7 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         ackHandle: string,
         trackerRefSeqNum: number, // back-compat summarizerNode - remove when fully enabled
         version?: IVersion,
+        summaryLogger: ITelemetryLogger = this.logger,
```
If there are not that many callers, I'd rather make it non-optional and make everybody pass it properly

vladsud commented on packages/runtime/container-runtime/src/summarizer.ts:
```diff
@@ -236,7 +239,8 @@ export class RunningSummarizer implements IDisposable {
         configuration: ISummaryConfiguration,
-        generateSummary: (full: boolean, safe: boolean) => Promise<GenerateSummaryData | undefined>,
+        // eslint-disable-next-line max-len
+        generateSummary: (full: boolean, safe: boolean, summaryLogger: ITelemetryLogger) => Promise<GenerateSummaryData | undefined>,
```
nit: here, and below: I think summaryLogger is too verbose. logger is good enough, there is no need to encode context (class name) into arg names

markfields replied:
summaryLogger isn't a class name - it's an important distinction, that the context of this logger is specific to the particular summar**y** (not summar**izer**).  I'm thinking about renaming this and the field in the logs to be about "runs", to leverage the terminology as in "RunningSummarizer".  Not sold on it though, still kind of verbose/potentially confusing.  #NamingIsHard

vladsud commented on packages/runtime/container-runtime/src/summarizer.ts:
```diff
@@ -442,6 +458,9 @@ export class RunningSummarizer implements IDisposable {
 
     private async summarizeCore(reason: string, safe: boolean): Promise<boolean | undefined> {
+        const summarizeCount = ++this.summarizeCount;
+        this.summaryLogger = new ChildLogger(this.summarizerLogger, "Summary", { summarizeCount });
```
Can we call them "baseLogger" and "currentLogger"?
summary vs. summarizer does not tell me much, I'd need to read the code to get this finer detail of what they mean

vladsud replied:
Also, have you considered to use single logger, but use property accessors (i.e. dynamic properties) as a way to set proper events? feels like it will be simpler, and code churn would be smaller...

arinwt replied:
The distinction is clear to me: Summarizer is the thing which summarizes, Summary is the summary being logged.

Dynamic property getters would be nice, but unfortunately they aren't as accurate when used in case of `refreshLatestSummaryAck`. There we need a way to verify that it is actually the same correlation (using summary sequence number or handle).
Maybe there's a way to get best of both worlds? Might be worth experimenting a little bit.

markfields replied:
Arin and I figured it out, thanks for the suggestion @vladsud

vladsud commented on packages/runtime/container-runtime/src/summarizer.ts:
```diff
@@ -728,7 +749,8 @@ export class Summarizer extends EventEmitter implements ISummarizer {
             this.configurationGetter(),
-            async (full: boolean, safe: boolean) => this.generateSummary(full, safe),
+            // eslint-disable-next-line max-len
+            async (full: boolean, safe: boolean, summaryLogger: ITelemetryLogger) => this.generateSummary(full, safe, summaryLogger),
```
same note about using bind()

markfields replied:
Will address on Monday

markfields committed:
Fix EsLint error

markfields committed:
Always use summaryLogger in generateSummary

markfields committed:
Fix state bug in summarizerNode

markfields committed:
Consolidate two loggers in RunningSummarizer

markfields committed:
lint

markfields commented:
Still some clean up to do, and I still want to look into testing.  I introduced a bug in it somewhere in my last couple commits that I just uncovered in manual testing...

markfields committed:
Bug fix and formatting

markfields committed:
Merge branch 'main' into summary-logger-id

markfields committed:
Merge branch 'main' into summary-logger-id

markfields committed:
More consistent/descriptive naming throughout

markfields committed:
Introduce ISummarizerInternalsProvider interface

This encapsulates some lambdas that were previously
passed around directly and hard to follow in the code.

markfields committed:
Build fix and tweaks

markfields committed:
rename log property

markfields committed:
Fix property name

markfields committed:
Enable "Debug currrent test" in container-runtime

markfields committed:
Add simple test for expected log sequence

markfields committed:
More specific log validation

markfields commented on packages/runtime/container-runtime/src/summarizer.ts:
```diff
@@ -500,9 +536,8 @@ export class RunningSummarizer implements IDisposable {
     private async generateSummaryWithLogging(message: string, safe: boolean): Promise<GenerateSummaryData | undefined> {
         const summarizingEvent = PerformanceEvent.start(this.logger, {
-            eventName: "Summarizing",
+            eventName: "SummarizerRun",
```
@arinwt, @vladsud, @anthony-murphy - heads up on this rename.

arinwt replied:
I think we should reconsider these names. (correct me if I'm wrong though)

My understanding is the new full event name will be: "... :Summarizer:Run:SummarizerRun".
Currently, when we "run" the summarizer, it means we begin the heuristics and start the RunningSummarizer. But here "run" means a specific summary is being created/tracked.  It seems like maybe we want to reconsider?

markfields replied:
Bah, good point about `Summarizer.run`. I was trying to get something unique to describe each iterative turn through the RunningSummarizer that generates a Summary.

How about................... `:Summarizer:Running:GenerateSummary_start`, and the property can be called `summaryGenTag`? Or maybe `summarizerRunTag` still works? Other ideas?

I want some other keyword (i.e. not just `summaryTag`) so we're not depending on distinguishing between Summarizing/Summarizer/Summary - I've found for myself and others that those words all run together in the brain.

arinwt replied:
That seems fine to me. I think there used to be another perf event within ContainerRuntime with a similar name, but since that was removed I agree that `GenerateSummary_*` is probably a good name.

Personally I like `summaryTag`, but yeah `summaryGenTag` is probably fine too.

anthony-murphy replied:
i like matching the method name in cases like this, as it makes look up in the code easy and obvious

markfields replied:
Thanks for the feedback 👍

markfields commented on packages/runtime/container-runtime/tsconfig.json:
```diff
@@ -8,7 +8,8 @@
         "types": [
-            "mocha"
+            "mocha",
+            "node"
```
Needed to run with "Debug current test" profile

arinwt replied:
I don't think we want to do this, since container-runtime package runs in browser. It doesn't actually hurt anything, but it makes it easier to accidentally use node-specific types.

Is there a way to only include the node-specific typings that are needed for the test in the test?
Even if it just means stubbing them locally?
FYI - @curtisman , @heliocliu

heliocliu replied:
You can also add a separate tsconfig file that extends the regular one and adds the types like this just for test; common-utils does this for its tests

markfields replied:
Ah very interesting.  I will revert, looks like we need to do some cleanup across the repo here, to get a tsconfig in the test dir everywhere so the launch config can use a consistent path to the test tsconfig.

markfields replied:
Opened #4397 for this.

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -1754,6 +1761,7 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
                     undefined,
                     version.id,
                     this.summaryTracker.referenceSequenceNumber,
+                    new ChildLogger(summaryLogger, undefined, { safeSummary: true }),
```
Future: I'd like to see the RunningSummarizer's logger be able to look up `safe` like it does `summarizerRunTag` so this won't be necessary but will be included automatically.

markfields commented on packages/runtime/container-runtime/src/test/summarizer.spec.ts:
```diff
@@ -4,24 +4,62 @@
 import { RunningSummarizer } from "../summarizer";
 import { SummaryCollection } from "../summaryCollection";
 
+class MockLogger extends TelemetryLogger implements ITelemetryLogger {
```
I will move this elsewhere in a future PR since it's general purpose.

markfields reviewed (commented):
Leaving some commentary for other reviewers

markfields committed:
Revert tsconfig change in this branch

markfields committed:
Fix tests

arinwt commented on packages/runtime/runtime-utils/src/summarizerNode.ts:
```diff
@@ -341,15 +347,22 @@ export class SummarizerNode implements ISummarizerNode {
         this.completeSummaryCore(proposalHandle, undefined, false);
     }
 
+    /**
+     * Recursive implementation for completeSummary, with additional internal-only parameters
+     */
```
Appreciate the extra comments!

markfields commented:
@arinwt Any idea why the "context reload is followed by an immediate summary" test would fail in CI but not locally for me?

arinwt reviewed (approved):
Looks good. I think some of the new event names are not my favorite, but will get used to them, and understand the drawbacks to the existing/alternatives.

arinwt commented:
> @arinwt Any idea why the "context reload is followed by an immediate summary" test would fail in CI but not locally for me?

Not off the top of my head.. I would guess maybe the merge with latest `main` broke it? But I don't know.

anthony-murphy commented:
@markfields seeing the same issue in my PR. running a build a master: https://dev.azure.com/fluidframework/public/_build/results?buildId=10160&view=results

anthony-murphy commented:
i think we both just broke the same test. they are super coupled to the string format of the logs. I filed an issue against these tests: #4415

markfields committed:
PR comments and test failure fix

markfields committed:
Try again to fix the contextReload test

My local repo is in a weird build state so hopefully this works...

markfields committed:
nits

markfields committed:
Merge branch 'main' into summary-logger-id

markfields committed:
merge/build break
